pkg/controller/atlasstream/atlasstream_instance_controller.go

```diff
@@ -0,0 +1,160 @@
+		return r.invalidate(isValid)
+	}
+
+	// check if stream instance is in "unsupported" state
+	if !r.AtlasProvider.IsResourceSupported(akoStreamInstance) {
+		return r.unsupport(workflowCtx)
+	}
+
+	project := akov2.AtlasProject{}
+	if err := r.Client.Get(ctx, akoStreamInstance.AtlasProjectObjectKey(), &project); err != nil {
+		return r.terminate(workflowCtx, workflow.Internal, err)
+	}
+
+	atlasClient, orgID, err := r.AtlasProvider.SdkClient(workflowCtx.Context, project.ConnectionSecretObjectKey(), log)
+	if err != nil {
+		return r.terminate(workflowCtx, workflow.AtlasAPIAccessNotConfigured, err)
+	}
+	workflowCtx.SdkClient = atlasClient
+	workflowCtx.OrgID = orgID
+
+	atlasStreamInstance, _, err := workflowCtx.SdkClient.StreamsApi.
+		GetStreamInstance(workflowCtx.Context, project.ID(), akoStreamInstance.Spec.Name).
+		Execute()
+
+	if err != nil && !admin.IsErrorCode(err, instanceNotFound) {
+		return r.terminate(workflowCtx, workflow.Internal, err)
+	}
+
+	isMarkedAsDeleted := !akoStreamInstance.GetDeletionTimestamp().IsZero()
+	isNotInAtlas := err != nil && admin.IsErrorCode(err, instanceNotFound)
+
+	switch {
+	case isNotInAtlas && !isMarkedAsDeleted:
+		// if no streams processing instance is not in atlas and is not marked as deleted - create
+		// hence, create the stream instance and transition to "ready" state
+		return r.create(workflowCtx, &project, akoStreamInstance)
+	case isMarkedAsDeleted:
+		// if a streams processing instance is marked as deleted,
+		// independently whether it exists in Atlas or not - delete
+		return r.delete(workflowCtx, &project, akoStreamInstance)
+	case hasChanged(akoStreamInstance, atlasStreamInstance):
+		// if a streams processing instance is ready and has changed - update
+		return r.update(workflowCtx, &project, akoStreamInstance)
+	}
+
+	// handle connection registry management
+	return r.handleConnectionRegistry(workflowCtx, &project, akoStreamInstance, atlasStreamInstance)
+}
+
+func hasChanged(streamInstance *akov2.AtlasStreamInstance, atlasStreamInstance *admin.StreamsTenant) bool {
+	config := streamInstance.Spec.Config
+	dataProcessRegion := atlasStreamInstance.GetDataProcessRegion()
+
+	return config.Provider != dataProcessRegion.GetCloudProvider() || config.Region != dataProcessRegion.GetRegion()
+}
+
+func (r *AtlasStreamsInstanceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("AtlasStreamInstance").
+		For(&akov2.AtlasStreamInstance{}, builder.WithPredicates(r.GlobalPredicates...)).
+		Watches(
+			&akov2.AtlasStreamConnection{},
+			handler.EnqueueRequestsFromMapFunc(r.findStreamInstancesForStreamConnection),
+		).
+		Complete(r)
+}
+
+func (r *AtlasStreamsInstanceReconciler) findStreamInstancesForStreamConnection(_ context.Context, obj client.Object) []reconcile.Request {
+	streamConnection, ok := obj.(*akov2.AtlasStreamConnection)
+	if !ok {
+		r.Log.Warnf("watching AtlasStreamConnection but got %t", obj)
```

Comment:
```suggestion
		r.Log.Warnf("watching AtlasStreamConnection but got %T", obj)
```

Note lowercase `%t` is for bools